Remotion/Data/DomainObjects/Mapping/Builder/ClassDefinitionBuilderNode.cs

```diff
@@ -25,32 +27,55 @@ namespace Remotion.Data.DomainObjects.Mapping.Builder
   /// Represents <see cref="Mapping.ClassDefinition"/> node in a tree constructed by <see cref="TypeDefinitionFactory"/>.
   /// Supports child navigation and correctly setting up child relationships in a resulting <see cref="Mapping.ClassDefinition"/>.
   /// </summary>
+  [DebuggerDisplay("Builder node for class '{Type}'")]
   public class ClassDefinitionBuilderNode : TypeDefinitionBuilderNode
   {
-    public ClassDefinitionBuilderNode? BaseClass { get; }
+    public ClassDefinitionBuilderNode? BaseClassNode { get; }
+
+    public IReadOnlyList<InterfaceDefinitionBuilderNode> ImplementedInterfaceNodes { get; }
+
+    [MemberNotNullWhen(false, nameof(BaseClassNode))]
+    public bool IsInheritanceRoot { get; }
 
     public ClassDefinition? ClassDefinition { get; private set; }
 
     public bool IsConstructed { get; private set; }
 
-    private readonly List<ClassDefinitionBuilderNode> _derivedClasses = new();
+    private readonly List<ClassDefinitionBuilderNode> _derivedClassNodes = new();
 
-    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClass)
+    public ClassDefinitionBuilderNode (Type type, ClassDefinitionBuilderNode? baseClassNode, IEnumerable<InterfaceDefinitionBuilderNode> implementedInterfaceNodes)
         : base(type)
     {
+      ArgumentUtility.CheckNotNull("implementedInterfaceNodes", implementedInterfaceNodes);
       if (!type.IsClass)
         throw new ArgumentException("The specified type must be a class.", "type");
-      if (baseClass?.IsConstructed == true)
-        throw new ArgumentException("The specified base class must not be a constructed node.", "baseClass");
-
-      BaseClass = baseClass;
-      BaseClass?.AddDerivedClass(this);
+      if (type == typeof(DomainObject))
+        throw new ArgumentException($"The specified type must not be the {nameof(DomainObject)} class.", "type");
+      if (!ReflectionUtility.IsDomainObject(type))
+        throw new ArgumentException($"The specified type must be a domain object.", "type");
```

Comment:
```suggestion
        throw new ArgumentException($"The specified type must be derived from '{nameof(DomainObject)}'.", "type");
```

Reply:
Interesting tidbit: I only saw the updated text first, not my review notes and the new messages where my first thoughts on the subject. So, I guess I'm consistent with my thoughts on messages and do seem to feel strongly about them :/